File: ios/Shared/CounterAPIClient.swift
import Foundation

actor CounterAPIClient {
    private let oauth: CounterOAuthClient
    private let session: URLSession
    private let sharedStore: CounterSharedStore?

    init(
        oauth: CounterOAuthClient,
        session: URLSession = .shared,
        sharedStore: CounterSharedStore? = CounterSharedStore()
    ) {
        self.oauth = oauth
        self.session = session
        self.sharedStore = sharedStore
    }

    func availableCounters(in environment: CounterEnvironment) async throws -> [CounterSelection] {
        let configuration = environment.oauthConfiguration
        let url = configuration.counterBaseURL.appending(path: "api/counter/available")
        let response: AvailableCountersResponse = try await request(url: url, method: "GET")
        let selections = response.counters.map {
            CounterSelection(
                organizationID: $0.organizationId,
                organizationName: $0.organizationName,
                counterID: $0.counterId,
                environment: environment
            )
        }

        if selections.count == 1, let selection = selections.first {
            try sharedStore?.saveSelection(selection)
        }

        return selections
    }

    func state(for selection: CounterSelection) async throws -> CounterSnapshot {
        let configuration = selection.environment.oauthConfiguration
        var components = URLComponents(
            url: configuration.counterBaseURL.appending(path: "api/counter/state"),
            resolvingAgainstBaseURL: false
        )
        components?.queryItems = [
            URLQueryItem(name: "organizationId", value: selection.organizationID),
            URLQueryItem(name: "counterId", value: selection.counterID),
        ]
        guard let url = components?.url else {
            throw CounterAPIError.invalidURL
        }

        let response: CounterStateResponse = try await request(url: url, method: "GET")
        let snapshot = CounterSnapshot(
            organizationID: selection.organizationID,
            counterID: selection.counterID,
            organizationName: selection.organizationName,
            count: response.count,
            updatedAt: .now
        )
        try sharedStore?.saveSnapshot(snapshot)
        return snapshot
    }

    func send(_ command: CounterCommand, for selection: CounterSelection) async throws -> CounterSnapshot {
        let configuration = selection.environment.oauthConfiguration
        let url = configuration.counterBaseURL.appending(path: "api/counter/command")
        let body = CounterCommandRequest(
            organizationId: selection.organizationID,
            counterId: selection.counterID,
            action: command.rawValue
        )
        let response: CounterStateResponse = try await request(
            url: url,
            method: "POST",
            body: try JSONEncoder().encode(body)
        )
        let snapshot = CounterSnapshot(
            organizationID: selection.organizationID,
            counterID: selection.counterID,
            organizationName: selection.organizationName,
            count: response.count,
            updatedAt: .now
        )
        try sharedStore?.saveSnapshot(snapshot)
        return snapshot
    }

    private func request<T: Decodable>(
        url: URL,
        method: String,
        body: Data? = nil
    ) async throws -> T {
        let accessToken = try await oauth.validAccessToken()
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw CounterAPIError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            let apiError = try? JSONDecoder().decode(CounterAPIErrorResponse.self, from: data)
            throw CounterAPIError.http(http.statusCode, apiError?.error)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

enum CounterCommand: String, Codable, Sendable {
    case increment
    case decrement
}

private struct CounterCommandRequest: Encodable {
    let organizationId: String
    let counterId: String
    let action: String
}

private struct CounterStateResponse: Decodable {
    let count: Int
}

private struct AvailableCountersResponse: Decodable {
    let counters: [AvailableCounterResponse]
}

private struct AvailableCounterResponse: Decodable {
    let organizationId: String
    let organizationName: String
    let counterId: String
}

private struct CounterAPIErrorResponse: Decodable {
    let error: String?
}

enum CounterAPIError: Error, LocalizedError {
    case invalidURL
    case invalidResponse
    case http(Int, String?)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Unable to build the Counter API URL."
        case .invalidResponse:
            return "The Counter API returned an invalid response."
        case .http(let status, let message):
            return message ?? "Counter API request failed with HTTP \(status)."
        }
    }
}

File: android/app/src/main/java/dev/niteowl/counter/api/CounterApiClient.kt
package dev.niteowl.counter.api

import dev.niteowl.counter.auth.OAuthClient
import dev.niteowl.counter.auth.OAuthConfig
import dev.niteowl.counter.data.CounterCommand
import dev.niteowl.counter.data.CounterEnvironment
import dev.niteowl.counter.data.CounterSelection
import dev.niteowl.counter.data.CounterSnapshot
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class CounterApiClient(private val oauth: OAuthClient) {
    suspend fun availableCounters(environment: CounterEnvironment): List<CounterSelection> {
        val config = OAuthConfig.forEnvironment(environment)
        val json = JSONObject(request(environment, "${config.counterBaseUrl}/api/counter/available", "GET"))
        val array: JSONArray = json.getJSONArray("counters")
        return buildList {
            for (index in 0 until array.length()) {
                val item = array.getJSONObject(index)
                add(
                    CounterSelection(
                        organizationId = item.getString("organizationId"),
                        organizationName = item.getString("organizationName"),
                        counterId = item.getString("counterId"),
                        environment = environment,
                    ),
                )
            }
        }
    }

    suspend fun state(selection: CounterSelection): CounterSnapshot {
        val config = OAuthConfig.forEnvironment(selection.environment)
        val url = "${config.counterBaseUrl}/api/counter/state?organizationId=${encode(selection.organizationId)}&counterId=${encode(selection.counterId)}"
        val json = JSONObject(request(selection.environment, url, "GET"))
        return snapshot(selection, json.getInt("count"))
    }

    suspend fun send(selection: CounterSelection, command: CounterCommand): CounterSnapshot {
        val config = OAuthConfig.forEnvironment(selection.environment)
        val body = JSONObject()
            .put("organizationId", selection.organizationId)
            .put("counterId", selection.counterId)
            .put("action", command.wireValue)
            .toString()
        val json = JSONObject(request(selection.environment, "${config.counterBaseUrl}/api/counter/command", "POST", body))
        return snapshot(selection, json.getInt("count"))
    }

    suspend fun registerFcmToken(selection: CounterSelection, token: String) {
        val config = OAuthConfig.forEnvironment(selection.environment)
        val body = JSONObject()
            .put("organizationId", selection.organizationId)
            .put("counterId", selection.counterId)
            .put("token", token)
            .toString()
        request(selection.environment, "${config.counterBaseUrl}/api/push/fcm", "POST", body)
    }

    private fun snapshot(selection: CounterSelection, count: Int) = CounterSnapshot(
        organizationId = selection.organizationId,
        organizationName = selection.organizationName,
        counterId = selection.counterId,
        count = count,
    )

    private suspend fun request(environment: CounterEnvironment, url: String, method: String, body: String? = null): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            setRequestProperty("Authorization", "Bearer ${oauth.validAccessToken(environment)}")
            setRequestProperty("Accept", "application/json")
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        if (body != null) connection.outputStream.use { it.write(body.toByteArray()) }
        val responseCode = connection.responseCode
        val response = (if (responseCode in 200..299) connection.inputStream else connection.errorStream)
            .bufferedReader().use { it.readText() }
        if (responseCode !in 200..299) error("Counter API request failed ($responseCode): $response")
        return response
    }

    private fun encode(value: String): String = URLEncoder.encode(value, Charsets.UTF_8)
}
